core/shared/src/main/scala/zio/stm/TSemaphore.scala

```diff
@@ -20,12 +20,13 @@ class TSemaphore private (val permits: TRef[Long]) extends AnyVal {
   final def acquire: STM[Nothing, Unit] = acquireN(1L)
 
   final def acquireN(n: Long): STM[Nothing, Unit] =
-    (for {
-      _     <- assertNonNegative(n)
-      value <- permits.get
-      _     <- STM.check(value >= n)
-      _     <- permits.set(value - n)
-    } yield ())
+    assertNonNegative(n).flatMap { _ =>
```

Comment:
You can probably save some typing by replacing `flatMap { _ => ... }` with `*>`, other than that it looks good.